DocumentsFromSnapshotMigration/src/test/java/com/rfs/EndToEndTest.java

```diff
@@ -0,0 +1,277 @@
+            var snapshotName = "my_snap";
+            var sourceClient = new OpenSearchClient(ConnectionContextTestParams.builder()
+                .host(sourceCluster.getUrl())
+                .insecure(true)
+                .build()
+                .toConnectionContext());
+            var snapshotCreator = new FileSystemSnapshotCreator(
+                snapshotName,
+                sourceClient,
+                SearchClusterContainer.CLUSTER_SNAPSHOT_DIR,
+                snapshotContext.createSnapshotCreateContext()
+            );
+            SnapshotRunner.runAndWaitForCompletion(snapshotCreator);
+            sourceCluster.copySnapshotData(localDirectory.toString());
+
+            var sourceRepo = new FileSystemRepo(localDirectory.toPath());
+            var targetClient = new OpenSearchClient(ConnectionContextTestParams.builder()
+                .host(targetCluster.getUrl())
+                .insecure(true)
+                .build()
+                .toConnectionContext());
+
+            // === ACTION: Migrate the templates and indices ===
+            migrateMetadata(
+                sourceRepo,
+                targetClient,
+                snapshotName,
+                List.of(templateName),
+                List.of(),
+                List.of(),
+                List.of(),
+                metadataContext,
+                sourceCluster.getVersion().getSourceVersion()
+            );
+
+            // Check that the templates were migrated
+            var targetClusterOperations = new ClusterOperations(targetCluster.getUrl());
+            var res = targetClusterOperations.get("/_template/" + templateName);
+            assertThat(res.getValue(), res.getKey(), equalTo(200));
+            assertThat(res.getValue(), Matchers.containsString("mappings\":{"));
+
+            // === ACTION: Migrate the documents ===
+            var runCounter = new AtomicInteger();
+            final var clockJitter = new Random(1);
+            var workerFuture = CompletableFuture.supplyAsync(
+                () -> migrateDocumentsSequentially(
+                    sourceRepo,
+                    snapshotName,
+                    List.of(),
+                    targetCluster.getUrl(),
+                    runCounter,
+                    clockJitter,
+                    testDocMigrationContext,
+                    sourceCluster.getVersion().getSourceVersion()
+                ),
+                executorService
+            );
+            try {
+                workerFuture.get();
+            } catch (ExecutionException e) {
+                var child = e.getCause();
+                if (child instanceof ExpectedMigrationWorkTerminationException) {
+                    // This is what we expect to happen
+                } else {
+                    throw Lombok.sneakyThrow(child);
+                }
+            } catch (InterruptedException e) {
+                throw Lombok.sneakyThrow(e);
+            }
```

Comment:
Since we aren't running the workers in parallel this can be greatly simplified down to something like the following:
```java
      var result = migrateDocumentsWithOneWorker(
          sourceRepo,
          snapshotName,
          List.of(),
          targetCluster.getUrl(),
          new AtomicInteger(),
          new Random(1),
          testDocMigrationContext,
          sourceCluster.getVersion().getSourceVersion()
      ));
      assertThat(result, equalTo(WORK_COMPLETED));
```

Reply:
Good callout.  I stuck with the more complex `migrateDocumentsSequentially` because there might be multiple shards worth of data and we want to ensure they are all migrated, but our final check that all our docs are migrated will catch if they aren't.  I'll move to the simpler approach and we can re-add the more complex one if/when required.

Reply:
Lets keep the test methods focused on what is being tested, all of these exception handling and unverified locals aren't relevant to this test scenario.  I'd be happy if you refactored it to work in parallel or with a single worker.